### PR TITLE
[Feature] Place and revert placing a candidate mutations

### LINKED ISSUE #9192: ✨ New graphql mutations for placing Candidates
## ✨ Feature

Create new specific mutations for placing (and un-placing) a Pool Candidate. 
Also add a some metadata to the Candidate model, like timestamp and the department they were placed in.

## 🌎 Context

These mutations will help us deprecate `updatePoolCandidateAsAdmin`, and hide the business logic of Pool Candidate state transitions in helpful ways (eg storing timestamps, and validating the transition to/from states). 

See parent issue #9149 for more context.

### New fields

Add two new fields to the PoolCandidate model, both protected by the `viewStatus` policy method:

- [ ] PoolCandidate.placed_at
  - default null 
  - in migration, default to now() if status implies qualification at some point: 
    - PLACED_CASUAL
    - PLACED_TERM
    - PLACED_INDETERMINATE
    - PLACED_TENTATIVE
- [ ] PoolCandidate.placed_department_id
  - default null
  - a foreign id on department table 
  - in schema, should be a relationship to a Department object, not the raw id

### Migrations

These migrations should all check the `PoolCandidatePolicy->updateStatus` policy method

- [ ] `placeCandidate({poolCandidateId!, placementType!, departmentId!})`
  - [ ] sets placedAt, a new field, to now()
  - [ ] placementType is an enum representing Tentative, Casual, Term, or Indeterminate placement
  - [ ] sets status to PLACED_TENTATIVE, PLACED_CASUAL, PLACED_TERM, PLACED_INDETERMINATE according to placementType
  - [ ] records placedDepartment, a new field
  - [ ] This is valid if candidate is QUALIFIED_AVAILABLE or currently PLACED. Throw an error otherwise.
- [ ] `unplaceCandidate({poolCandidateId!})`
  - **Question:** I'm open to naming suggestions on this 😆 
  - [ ] nullifies placedAt
  - [ ] nullifies placedDepartment
  - [ ] only valid if candidate currently in a PLACED_ status
  - [ ] sets status to QUALIFIED_AVAILABLE

## 🎨 Design File

This covers the Place and Un-place operations in this state diagram:
Pool Candidate State Transitions

### 🧑‍🎨 Designer

@tristan-orourke 

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the feature was properly implemented.

- [ ] New fields field added to database, model and schema
- [ ] The new fields have a value for existing rows where they should (see details)
- [ ] PoolCandidateFactory sets new fields appropriately if generating a candidate with a PLACED status
- [ ] New fields in schema require `viewStatus` policy to view

### New mutations:
- [ ] `placeCandidate({poolCandidateId!, placementType!, departmentId!})`
  - [ ] sets placedAt, a new field, to now()
  - [ ] placementType is an enum representing Tentative, Casual, Term, or Indeterminate placement
  - [ ] sets status to PLACED_TENTATIVE, PLACED_CASUAL, PLACED_TERM, PLACED_INDETERMINATE according to placementType
  - [ ] records placedDepartment, a new field
  - [ ] This is valid if candidate is QUALIFIED_AVAILABLE or currently PLACED (going from one placement to another is okay). Throw an error if coming from any other status.
- [ ] `unplaceCandidate({poolCandidateId!})`
  - **Question:** I'm open to naming suggestions on this 😆 
  - [ ] nullifies placedAt
  - [ ] nullifies placedDepartment
  - [ ] only valid if candidate currently in a PLACED_ status, throw an error otherwise
  - [ ] sets status to QUALIFIED_AVAILABLE
- [ ] New migrations all check the `PoolCandidatePolicy->updateStatus` policy method
- [ ] phpunit tests for new mutations, which check that status and timestamps are modified correctly, that applicants can't operate on their own applications, and that the mutations fail if coming from the wrong status.

### PR DESCRIPTION
**Pull request by vd1992**
🤖 Resolves #9192

## 👋 Introduction

Add the two new mutations.
Includes new fields, factory tweaks, and testing. The migration also updates some fields in certain cases. 

## 🧪 Testing

1. Migration runs right
2. Seeding fresh data is good
3. Test the two mutations confirming successful updates or errors depending on circumstances

Place a candidate

```
mutation placeCandidate {
  placeCandidate(
    id: "ID"
    placeCandidate: {
      placementType: PLACED_TERM, 
      departmentId: "ID"
    }
  ) {
    id
    status
    placedAt
    placedDepartment {
      id
    }
  }
}
```

Revert the above

```
mutation revertPlaceCandidate {
  revertPlaceCandidate(
    id: "ID"
  ) {
    id
    status
    placedAt
    placedDepartment {
      id
    }
  }
}
```

## 📸 Screenshot

image

image

image

### CODE REVIEW

File: api/app/GraphQL/Validators/PlaceCandidateValidator.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    public function rules(): array
+    {
+        $id = $this->arg('id');
+        $candidate = PoolCandidate::find($id);
+        $placedStatuses = array_column(PlacementType::cases(), 'name');
+        $statusesArray = [...$placedStatuses, PoolCandidateStatus::QUALIFIED_AVAILABLE->name];
+
+        if (! (in_array($candidate->pool_candidate_status, $statusesArray))) {
+            throw ValidationException::withMessages(['InvalidStatusForPlacing']);
```
Comment (esizer) [inline api/app/GraphQL/Validators/PlaceCandidateValidator.php:29]: Could we use the [`ApiErrorEnums`](https://github.com/GCTC-NTGC/gc-digital-talent/blob/main/api/database/helpers/ApiErrorEnums.php) for this?

Eventually these should be migrated to the new way of doing enums but for now I guess we can just use the existing file.

Comment (vd1992) [reply api/app/GraphQL/Validators/PlaceCandidateValidator.php]: Sure, I don't see why not. Done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/10013/commits/ea36a9c35918dc7a495672f62e041b1986a2c8dd

**Review by yonikid15 — COMMENTED** [review]
Nice work! Everything is working as expected 👍

Might be overkill but maybe add another php unit test to check switching from one placed status to another (eg. PLACED_TERM -> PLACED_CASUAL)?

Only comment I have is also getting approval for the name change to the mutation `unplaceCandidate`  -> `revertPlacedCandidate` (which is fine with me 😄) @tristan-orourke

**Comment by tristan-orourke** [comment]
I'm fine with the mutation name 👍

**Comment by vd1992** [comment]
Okay, who wants to do the honours? 😛 

image

**Review by yonikid15 — APPROVED** [review]

**Pull request merged**